### PR TITLE
CRIMAPP-683 add self employed client and partner

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change

Adds client and partner self employed journey.
Does not implement evidence prompt and task validation.

## Link to relevant ticket

[CRIMAPP-980](https://dsdmoj.atlassian.net/browse/CRIMAPP-980)
[CRIMAPP-981](https://dsdmoj.atlassian.net/browse/CRIMAPP-981)
[CRIMAPP-982](https://dsdmoj.atlassian.net/browse/CRIMAPP-982)
[CRIMAPP-986](https://dsdmoj.atlassian.net/browse/CRIMAPP-986)
[CRIMAPP-977](https://dsdmoj.atlassian.net/browse/CRIMAPP-977)
[CRIMAPP-1087](https://dsdmoj.atlassian.net/browse/CRIMAPP-1087)
[CRIMAPP-979](https://dsdmoj.atlassian.net/browse/CRIMAPP-979)
[CRIMAPP-979](https://dsdmoj.atlassian.net/browse/CRIMAPP-979)
[CRIMAPP-978](https://dsdmoj.atlassian.net/browse/CRIMAPP-978)
[CRIMAPP-1067](https://dsdmoj.atlassian.net/browse/CRIMAPP-1067)
[CRIMAPP-1065](https://dsdmoj.atlassian.net/browse/CRIMAPP-1065)

## Notes for reviewer

Uses single controller and forms for both client and partner.
Stores amount and frequency as jsonb within the model.

Extracted from https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/pull/1008

## How to manually test the feature

[CRIMAPP-980]: https://dsdmoj.atlassian.net/browse/CRIMAPP-980?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-981]: https://dsdmoj.atlassian.net/browse/CRIMAPP-981?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-982]: https://dsdmoj.atlassian.net/browse/CRIMAPP-982?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-986]: https://dsdmoj.atlassian.net/browse/CRIMAPP-986?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-977]: https://dsdmoj.atlassian.net/browse/CRIMAPP-977?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-1087]: https://dsdmoj.atlassian.net/browse/CRIMAPP-1087?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-979]: https://dsdmoj.atlassian.net/browse/CRIMAPP-979?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-979]: https://dsdmoj.atlassian.net/browse/CRIMAPP-979?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-978]: https://dsdmoj.atlassian.net/browse/CRIMAPP-978?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-1067]: https://dsdmoj.atlassian.net/browse/CRIMAPP-1067?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-1065]: https://dsdmoj.atlassian.net/browse/CRIMAPP-1065?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/models/business.rb [hunk 1]
```diff
@@ -1,3 +1,30 @@
 class Business < ApplicationRecord
   belongs_to :crime_application
+
+  store_accessor :address, :address_line_one, :address_line_two, :city, :country, :postcode
+
+  attribute :turnover, :amount_and_frequency
+  attribute :drawings, :amount_and_frequency
+  attribute :profit, :amount_and_frequency
+
```
Comment (timpeat) [inline app/models/business.rb:9]: Business#complete? is implemented in a forthcoming PR by @hiboabd

File: spec/forms/steps/income/business_employees_form_spec.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  let(:record) { Business.new }
+  let(:attributes) { {} }
+
+  it_behaves_like 'a form with a from_subject'
+
+  describe 'validations' do
+    it { is_expected.to validate_is_a(:has_employees, YesNoAnswer) }
+    it { is_expected.not_to validate_presence_of(:number_of_employees) }
+
+    context 'has additional owners' do
```
Comment (hiboabd) [inline spec/forms/steps/income/business_employees_form_spec.rb:17]: has employees

File: app/services/decisions/self_employed_income_decision_tree.rb [hunk 3]
```diff
@@ -1,7 +1,39 @@
 module Decisions
   class SelfEmployedIncomeDecisionTree < BaseDecisionTree
+    # TODO: consider renaming this decision tree business as it passes back to the income decision tree...
+
     # :nocov:
-    def destination; end
+    # temporary nocov until merge
```
Comment (hiboabd) [inline app/services/decisions/self_employed_income_decision_tree.rb:6]: Why is this necessary?

File: config/locales/en/errors.yml [hunk 4]
```diff
@@ -651,6 +657,67 @@ en:
...
+              blank: Enter the trading name of the business 
+            address:
+              address_line_one:
+                blank: Enter address line 1, typically the building and street
+              city:
+                blank: Enter town or city
+              country:
+                blank:  Enter country
+              postcode:
+                blank: Enter postcode
```
Comment (hiboabd) [inline config/locales/en/errors.yml:676]: Screenshot 2024-07-05 at 11 24 27
These error messages aren't appearing

**Review by hiboabd — COMMENTED** [review]
Looking good! Just spotted a few things. One other bug I spotted was with the income payments page, when you click save and continue without selecting any payments you get a error: `undefined method `work_benefits' for #<Steps::Income::IncomePaymentsForm:0x00000001148fee30` Doesn't seem to be happening on staging, so i assume it's a rebasing issue

File: app/presenters/summary/components/business.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+          ),
+          Components::FreeTextAnswer.new(
+            :number_of_employees, business.number_of_employees.to_s,
+            show: business.has_employees == 'yes'
+          ),
+          Components::AmountAndFrequencyAnswer.new(
+            :turnover, business.turnover, show: true
+          ),
+          Components::AmountAndFrequencyAnswer.new(
+            :profit, business.drawings, show: true
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/presenters/summary/components/business.rb:40]: 🦅 👁️  👍

**Review by hiboabd — APPROVED** [review]
Exciting! 🙌🚀
